app/models/git/exercise.rb

```diff
@@ -33,6 +33,17 @@ def synced_git_sha
       commit.oid
     end
 
+    def valid_submission_filepath?(filepath)
+      return false if filepath.match?(%r{[^a-zA-Z0-9_./-]})
+      return false if filepath.starts_with?(".meta")
+
+      # Some languages have solutions and tests in the same file so
+      # we need to protect for that here.
+      return false if test_filepaths.include?(filepath) && !solution_filepaths.include?(filepath)
+
+      true
```

Comment:
The if statement seems to cover the regular case whereas the comment describes the exceptional case

Reply:
The bit after `&&` is the bit that is referred to by the comment. I'll make it clearer.